pkg/workflows/ring/plugin.go

```diff
@@ -0,0 +1,290 @@
+		maxShardCount: cfg.MaxShardCount,
+		timeToSync:    cfg.TimeToSync,
+	}, nil
+}
+
+func (p *Plugin) Query(_ context.Context, _ ocr3types.OutcomeContext) (types.Query, error) {
+	return nil, nil
+}
+
+func (p *Plugin) Observation(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query) (types.Observation, error) {
+	shardHealth := p.store.GetShardHealth()
+
+	allWorkflowIDs := make([]string, 0)
+	for wfID := range p.store.GetAllRoutingState() {
+		allWorkflowIDs = append(allWorkflowIDs, wfID)
+	}
+	slices.Sort(allWorkflowIDs)
+
+	observation := &pb.Observation{
+		ShardHealthStatus: shardHealth,
+		WorkflowIDs:       allWorkflowIDs,
+		Now:               timestamppb.Now(),
+	}
+
+	return proto.MarshalOptions{Deterministic: true}.Marshal(observation)
+}
+
+func (p *Plugin) ValidateObservation(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, _ types.AttributedObservation) error {
+	return nil
+}
+
+func (p *Plugin) ObservationQuorum(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (quorumReached bool, err error) {
+	return quorumhelper.ObservationCountReachesObservationQuorum(quorumhelper.QuorumTwoFPlusOne, p.config.N, p.config.F, aos), nil
+}
+
+func (p *Plugin) Outcome(_ context.Context, outctx ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	// Load prior state
+	prior := &pb.Outcome{}
+	if outctx.PreviousOutcome == nil {
+		prior.Routes = map[string]*pb.WorkflowRoute{}
+		prior.State = &pb.RoutingState{Id: outctx.SeqNr, State: &pb.RoutingState_RoutableShards{RoutableShards: p.minShardCount}}
+	} else if err := proto.Unmarshal(outctx.PreviousOutcome, prior); err != nil {
+		return nil, err
+	}
+
+	currentShardHealth := make(map[uint32]int)
+	totalObservations := len(aos)
+	allWorkflows := make([]string, 0)
+	nows := make([]time.Time, 0)
+
+	// Collect shard health observations and workflows
+	for _, ao := range aos {
+		observation := &pb.Observation{}
+		if err := proto.Unmarshal(ao.Observation, observation); err != nil {
+			p.lggr.Warnf("failed to unmarshal observation: %v", err)
+			continue
+		}
+
+		for shardID, healthy := range observation.ShardHealthStatus {
+			if healthy {
+				currentShardHealth[shardID]++
+			}
+		}
+
+		// Collect workflow IDs
+		allWorkflows = append(allWorkflows, observation.WorkflowIDs...)
+
+		// Collect timestamps
+		if observation.Now != nil {
+			nows = append(nows, observation.Now.AsTime())
+		}
+	}
+
+	// Calculate median time
+	now := time.Now()
+	if len(nows) > 0 {
+		slices.SortFunc(nows, time.Time.Compare)
+		now = nows[len(nows)/2]
+	}
+
+	// Deduplicate workflows
+	workflowMap := make(map[string]bool)
+	for _, wf := range allWorkflows {
+		workflowMap[wf] = true
+	}
+	allWorkflows = make([]string, 0, len(workflowMap))
+	for wf := range workflowMap {
+		allWorkflows = append(allWorkflows, wf)
+	}
+	slices.Sort(allWorkflows) // Ensure deterministic order
```

Comment:
Iterating a map isn't great. Also, this takes n + n + n log n, it could be n + n

```
slices.Sort(allWorfklows)
dedupedWorkflows := make([]string, 0, len(allWorkflows)
var lastWorkflow string
for _, workflow := range allWorfklows {
  if lastWorkflow != workflow {
    lastWoflow = workflow
    dedupedWorkflows = append(dedupedWorkflows, workflow)
  }
}
```